### PR TITLE
Add support for the deprected and internal images list

### PR DESCRIPTION
**Pull request by wissamg**
## Status
Ready

## Related Content Pull Request
Related PR: 
[link to the PR at demisto/content](https://github.com/demisto/dockerfiles/pull/7666)

## Related Issues
Related: [CIAC-416](https://jira-hq.paloaltonetworks.local/browse/CIAC-416)

## Description
adding support for 2 couples of new key is the build.conf :
deprecated=true
deprectaed_reason="text message"
internalonly=true
internalonly_reason="text message"
the build script will handle the above key to maintains  2 list in the dockerfiles repo 
deprected_images.json: this file will contain all the deprecated docker images information
internal_images.json: this file will contain all the dockerimags information that are for internal use only

### CODE REVIEW

File: docker/add_image_to_deprecated_or_internal_list.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            if verbose:
+                print(image_list_string)
+            fp.write(image_list_string)
+            fp.close()
+            print(f"{image_name}: was added to the list {file_path}")
+            return error.entry_was_added
+        
+        except FileExistsError as exp:
+            print(f"{file_path}: is not exists make sure you are running from the the root folder of the dockerfiles repo (i.e.: /home/developer_name/dev/dockerfiles" \
+                "or make sure you are running the tools wiht fule file path to the file.")
```
Comment (Shellyber) [inline docker/add_image_to_deprecated_or_internal_list.py:70]:
```suggestion
                "or make sure you are running the tools with full file path to the file.")
```

File: docker/add_image_to_deprecated_or_internal_list.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            image_list_string = json.dumps(image_list)
+            if verbose:
+                print(image_list_string)
+            fp.write(image_list_string)
+            fp.close()
+            print(f"{image_name}: was added to the list {file_path}")
+            return error.entry_was_added
+        
+        except FileExistsError as exp:
+            print(f"{file_path}: is not exists make sure you are running from the the root folder of the dockerfiles repo (i.e.: /home/developer_name/dev/dockerfiles" \
```
Comment (Shellyber) [inline docker/add_image_to_deprecated_or_internal_list.py:69]:
```suggestion
            print(f"{file_path}: does not exists. make sure you are running from the the root folder of the dockerfiles repo (i.e.: /home/developer_name/dev/dockerfiles" \
```

File: docker/add_image_to_deprecated_or_internal_list.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    entry_exists = 1
+    reason_empty = 2
+    file_not_exists = 3
+    error_reading_file = 4
+    bad_json_file = 5
+    general_error = 6
+
+
+def add_image_to_deprecated_list(image_name :str, reason :str, file_path :str, verbose=False):
+    """ adding gioving docker image to the deprecation list. make sure no duplictation etc....
```
Comment (Shellyber) [inline docker/add_image_to_deprecated_or_internal_list.py:22]:
```suggestion
    """ adding giving docker image to the deprecation list. make sure no duplication etc....
```

File: docker/add_image_to_deprecated_or_internal_list.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            if not reason or len(reason) <= 0:
+                print ("reason field for the entry is empty")
+                return error.reason_empty
+
+            if(any (image_name in image["image_name"] for image in image_list)):
+                fp.close()
+                if verbose:
+                    image = [image for image in image_list if image_name in image['image_name']]
+                    print(image)
```
Comment (Shellyber) [inline docker/add_image_to_deprecated_or_internal_list.py:48]: Why do we need this print?

File: docker/build_docker.sh [hunk 5]
```diff
@@ -139,10 +139,35 @@ function docker_build {
...
+        if ! ${PY3CMD} "${DOCKER_SRC_DIR}"/add_image_to_deprecated_or_internal_list.py "${DOCKER_ORG}"/"${image_name}" "${reason}" "${DOCKER_SRC_DIR}"/deprecated_images.json; then
+            echo "failed while handling deprected image"
+            return 1
+        else
+            echo "deprectaed no need to build"
+            return 0
+        fi
+    fi
+
+    if [[ "$(prop 'internalonly')" ]]; then
```
Comment (Shellyber) [inline docker/build_docker.sh:161]: Let's remove this part in the meantime, until we decide how to handle this case of internal docker image

**Review by Shellyber — COMMENTED** [review]
Great work

File: docker/add_image_to_deprecated_or_internal_list.py (hunk 4 above)
Comment (Shellyber) [reply docker/add_image_to_deprecated_or_internal_list.py:48]: @wissamg Let's remove this print

**Review by Shellyber — APPROVED** [review]

**Comment by wissamg** [comment]
@glicht can you please review here or you would like us to have the conversation first.

**Comment by glicht** [comment]
@wissamg let's just add the env variables to the docker image when building a deprecated image:
* `DEPRECATED_IMAGE=true`
* `DEPRECATED_REASON=...`

File: README.md [hunk 6]
```diff
@@ -97,6 +97,8 @@ The build script will check for a `build.conf` file in the target image director
 
 * **version**: The version to use for tagging. Default: `1.0.0`. Note: that additionally, the CircleCI build number is always appended to the version as a revision (for example: `1.0.0.15519`) to create a unique version per build.
 * **devonly**: If set the image will be pushed only to the `devdemisto` org in docker hub and will not be pushed to the `demisto` org. Should be used for images which are for development purposes only (such as the image used in CircleCI to build this project).
+* **deprecated**: If set the image will be listed as deprecated in the deprecated_images.json file and the image will be forbidden form using in the integrations/automations.
+* **deprecated-reason**: Free text that explain the deprecation reason.
```
Comment (glicht) [inline README.md:101]: Isn't this supposed to be `deprecated_reasons` ?

File: README.md [hunk 7]
```diff
@@ -205,3 +207,24 @@ For example:
 ```
 ./docker/add_dependabot.sh docker/nmap
 ```
+
+### How to mark image as deprecated
+
+To mark an image as deprecated please follow the following steps:
+
+* 1- Add the following tow keys to the build.conf of the image.
```
Comment (glicht) [inline README.md:215]:
```suggestion
1. Add the following two keys to the build.conf of the image:
```

File: README.md [hunk 8]
```diff
@@ -205,3 +207,24 @@ For example:
...
+
+  * deprecated=true
+  * deprecated_reason="free text"
```
Comment (glicht) [inline README.md:218]: Do we need the quotes? Would be better we didn't have quotes.

File: docker/build_docker.sh [hunk 9]
```diff
@@ -139,10 +139,24 @@ function docker_build {
...
+
+    if [[ "$(prop 'deprecated')" ]]; then
+        echo "${DOCKER_ORG}/${image_name} image is deprected, checking whethear the image is listed in the deprecated list or not"
+        reason=$(prop 'deprecated_reason')
+        
+        if ! ${PY3CMD} "${DOCKER_SRC_DIR}"/add_image_to_deprecated_or_internal_list.py "${DOCKER_ORG}"/"${image_name}" "${reason}" "${DOCKER_SRC_DIR}"/deprecated_images.json; then
+            echo "failed while handling deprected image"
+            return 1
+        else
+            echo "deprectaed no need to build"
```
Comment (glicht) [inline docker/build_docker.sh:156]: I think this message is not accurate. We build anyway even if it is deprecated (which is ok). But let's just not have a confusing mesasge.

File: docker/build_docker.sh [hunk 10]
```diff
@@ -159,6 +173,13 @@ function docker_build {
     cp Dockerfile "$tmp_dir/Dockerfile"
     echo "" >> "$tmp_dir/Dockerfile"
     echo "ENV DOCKER_IMAGE=$image_full_name" >> "$tmp_dir/Dockerfile"
+    
+    if [[ "$(prop 'deprecated')" ]]; then
+        echo "DEPRECATED_IMAGE=true" >> "$tmp_dir/Dockerfile"
```
Comment (glicht) [inline docker/build_docker.sh:178]: I think you need `ENV`

```suggestion
        echo "ENV DEPRECATED_IMAGE=true" >> "$tmp_dir/Dockerfile"
```

File: docker/build_docker.sh [hunk 11]
```diff
@@ -159,6 +173,13 @@ function docker_build {
...
+        reason=$(prop 'deprecated_reason')
+        echo "DEPRECATED_REASON=$reason"
```
Comment (glicht) [inline docker/build_docker.sh:180]:
```suggestion
        echo "ENV DEPRECATED_REASON=$reason"
```

**Review by glicht — CHANGES_REQUESTED** [review]
Please see my comments

File: README.md (hunk 6 above)
Comment (wissamg) [reply README.md:101]: yes you are right

**Review by glicht — APPROVED** [review]

**Pull request merged**